google-ads-examples/src/main/java/com/google/ads/googleads/examples/remarketing/UploadConversionEnhancement.java

```diff
@@ -0,0 +1,282 @@
+              googleAdsClient,
+              params.customerId,
+              params.conversionActionId,
+              params.orderId,
+              params.userAgent,
+              params.restatementValue,
+              params.currencyCode);
+    } catch (GoogleAdsException gae) {
+      // GoogleAdsException is the base class for most exceptions thrown by an API request.
+      // Instances of this exception have a message and a GoogleAdsFailure that contains a
+      // collection of GoogleAdsErrors that indicate the underlying causes of the
+      // GoogleAdsException.
+      System.err.printf(
+          "Request ID %s failed due to GoogleAdsException. Underlying errors:%n",
+          gae.getRequestId());
+      int i = 0;
+      for (GoogleAdsError googleAdsError : gae.getGoogleAdsFailure().getErrorsList()) {
+        System.err.printf("  Error %d: %s%n", i++, googleAdsError);
+      }
+      System.exit(1);
+    }
+  }
+
+  /**
+   * Runs the example.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   * @param conversionActionId conversion action ID associated with this conversion.
+   * @param orderId unique order ID (transaction ID) of the conversion.
+   * @param userAgent the HTTP user agent of the conversion.
+   * @param restatementValue the enhancement value
+   * @param restatementCurrencyCode the currency of the enhancement value
+   */
+  // [START upload_conversion_enhancement]
+  private void runExample(
+      GoogleAdsClient googleAdsClient,
+      long customerId,
+      long conversionActionId,
+      String orderId,
+      String userAgent,
+      Double restatementValue,
+      String restatementCurrencyCode)
+      throws NoSuchAlgorithmException, UnsupportedEncodingException {
+    // [START create_adjustment]
+    // Creates a builder for constructing the enhancement adjustment.
+    ConversionAdjustment.Builder enhancementBuilder =
+        ConversionAdjustment.newBuilder()
+            .setConversionAction(ResourceNames.conversionAction(customerId, conversionActionId))
+            .setAdjustmentType(ConversionAdjustmentType.ENHANCEMENT)
+            // Enhancements MUST use order ID instead of GCLID date/time pair.
+            .setOrderId(orderId);
+
+    // Creates a SHA256 message digest for hashing user identifiers in a privacy-safe way, as
+    // described at https://support.google.com/google-ads/answer/9888656.
+    MessageDigest sha256Digest = MessageDigest.getInstance("SHA-256");
+
+    // Adds user identifiers, hashing where required.
+
+    // Creates a user identifier using sample values for the user address.
+    UserIdentifier addressIdentifier =
+        UserIdentifier.newBuilder()
+            // User identifiers for conversion enhancements MUST use first party data.
+            .setUserIdentifierSource(UserIdentifierSource.FIRST_PARTY)
+            .setAddressInfo(
+                OfflineUserAddressInfo.newBuilder()
```

Comment:
It doesn't look like the OfflineUserAddressInfo builder is ever actually built. Is this correct? Does it work?